### PR TITLE
[1LP][RFR] Convert AnalysisProfile to Widgetastic

### PR DESCRIPTION
**Pull request by mshriver**
Separate AnalysisProfile into its own module
Widgetastic conversion of the pages and CRUD

### Included tests
All tests in `cfme/tests/configure/test_analysis_profiles.py` should pass.

NOT Including test_compliance.py and test_instance_analysis.py because I slightly changed the yield_fixture there for an AnalysisProfile object, and the tests pretty much completely failed for unrelated reasons.

{{ pytest: -v --long-running cfme/tests/configure/test_analysis_profiles.py }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit c2451167a388648fe1aacfbbd174f553c0b2a993:

`cfme/configure/configuration/analysis_profile.py`:
- :red_circle: Line 2:1: **F401** *'navmazing.NavigateToAttribute' imported but unused*
- :red_circle: Line 5:1: **F401** *'widgetastic_patternfly.Button' imported but unused*
- :red_circle: Line 5:1: **F401** *'widgetastic_patternfly.FlashMessages' imported but unused*
- :red_circle: Line 6:1: **F401** *'widgetastic_manageiq.ItemsToolBarViewSelector' imported but unused*
- :red_circle: Line 6:1: **F401** *'widgetastic_manageiq.Accordion' imported but unused*
- :red_circle: Line 6:1: **F401** *'widgetastic_manageiq.TimelinesView' imported but unused*
- :red_circle: Line 6:1: **F401** *'widgetastic_manageiq.Search' imported but unused*
- :red_circle: Line 6:1: **F401** *'widgetastic_manageiq.SummaryTable' imported but unused*
- :red_circle: Line 6:1: **F401** *'widgetastic_manageiq.ManageIQTree' imported but unused*
- :red_circle: Line 11:1: **F401** *'cfme.base.ui.Region' imported but unused*
- :red_circle: Line 12:1: **F401** *'utils.appliance.current_appliance' imported but unused*
- :red_circle: Line 15:1: **F401** *'utils.timeutil.parsetime' imported but unused*
- :red_circle: Line 17:1: **F401** *'utils.wait.wait_for' imported but unused*
- :bangbang: Line 139:9: **E265** *block comment should start with '# '*
- :bangbang: Line 140:9: **E265** *block comment should start with '# '*
- :red_circle: Line 145:13: **F821** *undefined name 'fill'*
- :red_circle: Line 145:42: **F821** *undefined name 'form_buttons'*
- :red_circle: Line 147:13: **F821** *undefined name 'fill'*
- :red_circle: Line 147:45: **F821** *undefined name 'form_buttons'*
- :red_circle: Line 148:9: **F821** *undefined name 'flash'*
- :red_circle: Line 152:9: **F821** *undefined name 'tb'*
- :red_circle: Line 153:9: **F821** *undefined name 'sel'*
- :red_circle: Line 154:9: **F821** *undefined name 'flash'*
- :red_circle: Line 162:9: **F821** *undefined name 'fill'*
- :red_circle: Line 163:21: **F821** *undefined name 'form_buttons'*
- :red_circle: Line 164:9: **F821** *undefined name 'flash'*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/configure/configuration/analysis_profile.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    def exists(self):
+        try:
+            navigate_to(self, 'Details')
+        except (NavigationDestinationNotFound, CandidateNotFound):
+            return False
+        else:
+            return True
+
+    def __str__(self):
+        return self.name
```
Comment (mfalesni) [inline cfme/configure/configuration/analysis_profile.py]: You should probably also add Fillable mixin and implement as_fill_value with the same functionality, se we can then remove the `__str__` once we get rid of old style models.

Comment (mshriver) [inline cfme/configure/configuration/analysis_profile.py]: Adding

File: widgetastic_manageiq/__init__.py [hunk 2]
```diff
@@ -1957,3 +1962,37 @@ def __str__(self):
...
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.action_row = kwargs.pop('action_row', 0)  # pull this off and pass the rest up
+        super(DynamicTable, self).__init__(*args, **kwargs)
+
+    def row_add(self):
+        """Use the action-cell column widget to add a row"""
+        try:
+            self.__getitem__(self.action_row).click()
```
Comment (mfalesni) [inline widgetastic_manageiq/__init__.py]: `__getitem__(x)` == `[x]`

File: cfme/configure/configuration/analysis_profile.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    def form_fill_args(self, updates=None):
+        """Build a dictionary of nested tab_forms for assoc_fill from a flat object dictionary
+        If updates dictionary is passed, it is used instead of `self`
+        """
+        # well this is ugly. but works for create and update
+        # when updates is dict, getattr results in errors, using the default breaks update
+        # when updates is None, calling get results in an error because its None.get
+        # TODO FIXME
+        # For some reason the dict copy doesn't seem to be working here? its updating self.attr
+        # when popping the assoc_column item instead of the new dict.
```
Comment (mfalesni) [inline cfme/configure/configuration/analysis_profile.py]: `from copy import copy; a = copy(b);` ?

Comment (mshriver) [inline cfme/configure/configuration/analysis_profile.py]: I'll toy with copying the entire object instead of just copying the its `__dict__`.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e830f8455f886d440cbbea83cb5ede3cf274b525:

`cfme/configure/configuration/analysis_profile.py`:
- :red_circle: Line 2:1: **F401** *'functools.partial' imported but unused*
- :red_circle: Line 3:1: **F401** *'copy.copy' imported but unused*
- :red_circle: Line 302:5: **F811** *redefinition of unused 'copy' from line 3*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ce4486f4a72f4d0cd185c4b63ca1c0f8b05dbc60:

`cfme/configure/configuration/analysis_profile.py`:
- :red_circle: Line 2:1: **F401** *'functools.partial' imported but unused*
- :red_circle: Line 3:1: **F401** *'copy.copy' imported but unused*
- :red_circle: Line 302:5: **F811** *redefinition of unused 'copy' from line 3*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 08c953f5cb4e235d60c8537f8dd7106d0ef7dfa9:

`cfme/configure/configuration/analysis_profile.py`:
- :red_circle: Line 2:1: **F401** *'functools.partial' imported but unused*
- :red_circle: Line 3:1: **F401** *'copy.copy' imported but unused*
- :red_circle: Line 302:5: **F811** *redefinition of unused 'copy' from line 3*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f59737591a800c05b837beb883c5113159f86efc:

`cfme/tests/control/test_compliance.py`:
- :red_circle: Line 12:1: **F401** *'cfme.web_ui.flash' imported but unused*
- :red_circle: Line 12:1: **F401** *'cfme.web_ui.toolbar' imported but unused*
- :red_circle: Line 14:1: **F401** *'utils.version' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/configure/configuration/analysis_profile.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+class AnalysisProfileAddView(BaseLoggedInPage):
+    """View for the add form, switches between host/vm based on object type
+    Uses a switchable view based on the profile type widget
+    """
+    @property
+    def is_displayed(self):
+        return (
+            self.title.text == 'Adding a new Analysis Profile' and
+            self.profile_type.text == self.context['object'].profile_type)
+    title = Text('//div[@id="main-content"]//h1[@id="explorer_title"]'
```
Comment (quarckster) [inline cfme/configure/configuration/analysis_profile.py]: **REQUIRED**
Please put one blank line between title and is_displayed.

File: cfme/configure/configuration/analysis_profile.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+        if cancel:
+            view.cancel.click()
+        else:
+            view.add.click()
+
+        view = self.create_view(AnalysisProfileAllView)
+        view.flush_widget_cache()
+        assert view.is_displayed
+        view.flash.assert_no_error()
```
Comment (quarckster) [inline cfme/configure/configuration/analysis_profile.py]: Not needed

File: cfme/configure/configuration/analysis_profile.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+        form_fill_args = self.form_fill_args(updates=updates)
+        view = navigate_to(self, 'Edit')
+        changed = view.form.fill(form_fill_args)
+
+        if changed and not cancel:  # save button won't be enabled if nothing was changed
+            view.save.click()
+        else:
+            view.cancel.click()
+
+        view = self.create_view(AnalysisProfileDetailsView)  # edited through details
```
Comment (quarckster) [inline cfme/configure/configuration/analysis_profile.py]: **REQUIRED**
```python
view = self.create_view(AnalysisProfileDetailsView, override=updates)
```

File: cfme/configure/configuration/analysis_profile.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+        view.flush_widget_cache()
+        # Can't assert view.is_displayed because the name may have changed and object hasn't
+        # been updated yet
+        assert (view.entities.title.text == 'Settings Analysis Profile "{}"'
+                                            .format(updates.get('name', self.name)))
+        view.flash.assert_no_error()
```
Comment (quarckster) [inline cfme/configure/configuration/analysis_profile.py]: Not needed

File: cfme/configure/configuration/analysis_profile.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+            view.cancel.click()
+        else:
+            view.add.click()
+
+        # check the result
+        view = self.create_view(
+            AnalysisProfileDetailsView if cancel else AnalysisProfileAllView)
+        view.flush_widget_cache()
+        assert view.is_displayed
+        view.flash.assert_no_error()
```
Comment (quarckster) [inline cfme/configure/configuration/analysis_profile.py]: Not needed

File: cfme/configure/configuration/analysis_profile.py (hunk 6 above)
Comment (mshriver) [reply cfme/configure/configuration/analysis_profile.py]: Testing

Comment (mshriver) [reply cfme/configure/configuration/analysis_profile.py]: Looks like it's working now - when I had tried before I think I had other bugs preventing it from working correctly.

Awesome! Thanks @quarckster

**Pull request merged**